File: decx-cli/src/core/installer.ts
/**
 * decx-server.jar finder and installer.
 *
 * Release discovery deliberately avoids the GitHub REST API, which is
 * rate-limited to 60 requests/hour per IP for unauthenticated clients:
 * - the latest stable version comes from the npm registry (`@jygzyc/decx-cli`
 *   is published from the same tag as the server jar)
 * - prerelease versions come from the GitHub releases atom feed
 * - jar assets are downloaded from deterministic /releases/download URLs
 * None of these require the GitHub API, a token, or an installed `gh` CLI.
 */

import * as path from "path";
import { existsSync, mkdirSync, readFileSync, renameSync, unlinkSync } from "fs";
import { inflateRawSync } from "node:zlib";
import { downloadWithProgress } from "../utils/progress.js";
import { decxPath } from "./paths.js";

const DECX_SERVER_HOME: string | undefined = process.env.DECX_SERVER_HOME;
const DEFAULT_FETCH = fetch;

const NPM_PACKAGE = "@jygzyc/decx-cli";
const NPM_LATEST_URL = `https://registry.npmjs.org/${NPM_PACKAGE}/latest`;
const GITHUB_REPO = "jygzyc/decx";
const RELEASES_ATOM_URL = `https://github.com/${GITHUB_REPO}/releases.atom`;

/**
 * Compare two semver strings (e.g. "2.2.1" vs "2.3.0").
 * Returns >0 if a > b, <0 if a < b, 0 if equal.
 */
export function compareSemver(a: string, b: string): number {
  const pa = a.split("-")[0].split(".").map(Number);
  const pb = b.split("-")[0].split(".").map(Number);
  for (let i = 0; i < 3; i++) {
    const va = pa[i] ?? 0;
    const vb = pb[i] ?? 0;
    if (va !== vb) return va - vb;
  }
  return 0;
}

const INSTALL_DIR = decxPath("bin");
const INSTALL_PATH = path.join(INSTALL_DIR, "decx-server.jar");

export interface ReleaseAsset {
  name: string;
  browser_download_url: string;
}

interface ReleaseSummary {
  tag_name: string;
  assets: ReleaseAsset[];
}

export type InstallDecxServerResult =
  | { ok: true; message: string; version: string; path: string }
  | { ok: false; message: string };

interface InstallDecxServerOptions {
  fetchImpl?: typeof fetch;
  downloadWithProgressImpl?: typeof downloadWithProgress;
  installDir?: string;
  installPath?: string;
  logger?: Pick<Console, "error">;
  /** Installed server version; when it matches the latest release, skip the download. */
  currentVersion?: string;
}

type ReleaseFetchResult =
  | { ok: true; release: ReleaseSummary }
  | { ok: false; message: string };

function assetUrl(version: string): string {
  return `https://github.com/${GITHUB_REPO}/releases/download/v${version}/decx-server-${version}.jar`;
}

/** Build a release summary from a version using deterministic asset naming. */
function summaryForVersion(version: string): ReleaseSummary {
  return {
    tag_name: `v${version}`,
    assets: [{ name: `decx-server-${version}.jar`, browser_download_url: assetUrl(version) }],
  };
}

/** Matches version tags with a prerelease suffix, e.g. v4.2.0-rc.1. */
const PRERELEASE_VERSION_RE = /^v?\d+\.\d+\.\d+-/;

/**
 * Discover the latest stable release (npm registry) or newest prerelease
 * (GitHub releases atom feed). Both sources are unauthenticated and not
 * subject to GitHub API rate limits.
 */
async function fetchReleaseSummary(
  prerelease: boolean,
  fetchImpl: typeof fetch = DEFAULT_FETCH,
  timeoutMs: number = 15_000,
): Promise<ReleaseFetchResult> {
  const signal = AbortSignal.timeout(timeoutMs);

  if (prerelease) {
    try {
      const res = await fetchImpl(RELEASES_ATOM_URL, {
        headers: { Accept: "application/atom+xml" },
        signal,
      });
      if (res.ok) {
        const text = await res.text();
        const titles = [...text.matchAll(/<title>([^<]+)<\/title>/g)].map((m) => m[1].trim());
        const pre = titles.find((title) => PRERELEASE_VERSION_RE.test(title));
        if (pre) {
          return { ok: true, release: summaryForVersion(pre.replace(/^v/, "")) };
        }
      }
      return { ok: false, message: "No prerelease found" };
    } catch (err) {
      return {
        ok: false,
        message: `Failed to reach GitHub releases feed: ${err instanceof Error ? err.message : String(err)}`,
      };
    }
  }

  try {
    const res = await fetchImpl(NPM_LATEST_URL, {
      headers: { Accept: "application/json" },
      signal,
    });
    if (res.ok) {
      const data = await res.json() as { version?: string };
      if (typeof data.version === "string" && data.version.length > 0) {
        return { ok: true, release: summaryForVersion(data.version) };
      }
    }
    return {
      ok: false,
      message: `Failed to fetch latest version from the npm registry (HTTP ${res.status})`,
    };
  } catch (err) {
    return {
      ok: false,
      message: `Failed to reach npm registry: ${err instanceof Error ? err.message : String(err)}`,
    };
  }
}

/**
 * Latest stable version string from the npm registry, or null when the
 * lookup fails (network error, non-OK response, malformed payload).
 */
export async function fetchLatestStableVersion(
  fetchImpl: typeof fetch = DEFAULT_FETCH,
  timeoutMs: number = 15_000,
): Promise<string | null> {
  const fetched = await fetchReleaseSummary(false, fetchImpl, timeoutMs);
  return fetched.ok ? normalizeVersion(fetched.release.tag_name) : null;
}

/**
 * Find decx-server.jar from known locations.
 * Priority: DECX_SERVER_HOME env > ~/.decx/bin/decx-server.jar
 */
export function findDecxServerJar(): string | null {
  if (DECX_SERVER_HOME) {
    if (DECX_SERVER_HOME.endsWith(".jar") && existsSync(DECX_SERVER_HOME)) return DECX_SERVER_HOME;
    const fromDir = path.join(DECX_SERVER_HOME, "decx-server.jar");
    if (existsSync(fromDir)) return fromDir;
  }

  if (existsSync(INSTALL_PATH)) return INSTALL_PATH;

  return null;
}

export function selectDecxServerAsset(assets: ReleaseAsset[]): ReleaseAsset | undefined {
  return assets.find((asset) => asset.name.includes("decx-server") && asset.name.endsWith(".jar"));
}

function normalizeVersion(tag: string): string {
  return tag.replace(/^v/, "");
}

function replaceInstalledJar(tmpPath: string, installPath: string): void {
  const backupPath = `${installPath}.bak`;
  const hadExisting = existsSync(installPath);

  try {
    if (hadExisting) {
      renameSync(installPath, backupPath);
    }
    renameSync(tmpPath, installPath);
    if (hadExisting && existsSync(backupPath)) {
      unlinkSync(backupPath);
    }
  } catch (error) {
    // Best-effort cleanup: drop the partial download and restore the old jar.
    if (existsSync(tmpPath)) {
      try {
        unlinkSync(tmpPath);
      } catch { /* ignore */ }
    }
    if (hadExisting && !existsSync(installPath) && existsSync(backupPath)) {
      try {
        renameSync(backupPath, installPath);
      } catch { /* ignore */ }
    }
    throw error;
  }
}

/**
 * Check if a newer server version is available.
 * `error` is set when the check itself failed (network/API error) — callers
 * must not treat that as "no update available".
 */
export async function checkForServerUpdate(
  currentVersion: string,
  prerelease: boolean = false,
  options: { fetchImpl?: typeof fetch } = {},
): Promise<{ available: boolean; latestVersion: string; error?: string }> {
  const fetched = await fetchReleaseSummary(prerelease, options.fetchImpl ?? DEFAULT_FETCH);
  if (!fetched.ok) {
    return { available: false, latestVersion: currentVersion, error: fetched.message };
  }

  const latest = normalizeVersion(fetched.release.tag_name);
  const available = compareSemver(latest, currentVersion.replace(/^v/, "")) > 0;
  return { available, latestVersion: latest };
}

/**
 * Read `version=<x.y.z>` from the `version.properties` entry inside a jar
 * (the same resource `DecxConstants` reads at server startup).
 *
 * Parses the zip central directory directly — no external zip dependency —
 * and supports both stored and deflated entries. Returns null when the file
 * is missing, malformed, or has no readable version.
 */
export function readJarVersionProperty(jarPath: string): string | null {
  let buf: Buffer;
  try {
    buf = readFileSync(jarPath);
  } catch {
    return null;
  }
  try {
    const eocd = buf.lastIndexOf(Buffer.from([0x50, 0x4b, 0x05, 0x06])); // "PK\x05\x06"
    if (eocd < 0) return null;
    const entryCount = buf.readUInt16LE(eocd + 10);
    let p = buf.readUInt32LE(eocd + 16); // central directory offset
    for (let i = 0; i < entryCount; i++) {
      if (buf.readUInt32LE(p) !== 0x02014b50) return null; // "PK\x01\x02"
      const method = buf.readUInt16LE(p + 10);
      const compressedSize = buf.readUInt32LE(p + 20);
      const nameLen = buf.readUInt16LE(p + 28);
      const extraLen = buf.readUInt16LE(p + 30);
      const commentLen = buf.readUInt16LE(p + 32);
      const localHeaderOffset = buf.readUInt32LE(p + 42);
      const name = buf.toString("utf8", p + 46, p + 46 + nameLen);
      if (name === "version.properties") {
        // Local header repeats name/extra lengths; data follows them.
        const lNameLen = buf.readUInt16LE(localHeaderOffset + 26);
        const lExtraLen = buf.readUInt16LE(localHeaderOffset + 28);
        const dataStart = localHeaderOffset + 30 + lNameLen + lExtraLen;
        const data = buf.subarray(dataStart, dataStart + compressedSize);
        const content = method === 8 ? inflateRawSync(data).toString("utf8") : data.toString("utf8");
        const match = content.match(/^version=([\w.-]+)\s*$/m);
        return match ? match[1] : null;
      }
      p += 46 + nameLen + extraLen + commentLen;
    }
  } catch {
    return null;
  }
  return null;
}

/** Version recorded inside the installed decx-server.jar, or null when absent/unreadable. */
export function installedServerVersion(installPath: string = INSTALL_PATH): string | null {
  return existsSync(installPath) ? readJarVersionProperty(installPath) : null;
}

/**
 * Download and install the latest decx-server.jar from GitHub releases.
 */
export async function installDecxServer(
  prerelease: boolean = false,
  options: InstallDecxServerOptions = {}
): Promise<InstallDecxServerResult> {
  const {
    fetchImpl = DEFAULT_FETCH,
    downloadWithProgressImpl = downloadWithProgress,
    installDir = INSTALL_DIR,
    installPath = INSTALL_PATH,
    logger = console,
    currentVersion,
  } = options;

  try {
    logger.error(`  Fetching latest ${prerelease ? "prerelease" : "release"} info...`);

    const fetched = await fetchReleaseSummary(prerelease, fetchImpl);
    if (!fetched.ok) {
      return { ok: false, message: fetched.message };
    }
    const release = fetched.release;

    const version = normalizeVersion(release.tag_name);

    // Prefer the version recorded inside the installed jar itself (ground
    // truth) over the config record, which may be missing or stale after a
    // manual jar replacement. Local and remote versions match: nothing to
    // download.
    const jarVersion = existsSync(installPath) ? readJarVersionProperty(installPath) : null;
    const effectiveVersion = jarVersion ?? currentVersion;
    if (effectiveVersion !== undefined && effectiveVersion === version && existsSync(installPath)) {
      return {
        ok: true,
        message: `decx-server is already up to date (v${version})`,
        version,
        path: installPath,
      };
    }

    const asset = selectDecxServerAsset(release.assets);

    if (!asset) {
      return { ok: false, message: `No decx-server jar asset found in release ${release.tag_name}` };
    }

    mkdirSync(installDir, { recursive: true });

    const downloadRes = await fetchImpl(asset.browser_download_url, { redirect: "follow" });
    if (!downloadRes.ok || !downloadRes.body) {
      return {
        ok: false,
        message: `Download failed: HTTP ${downloadRes.status} (the release may still be publishing; retry shortly)`,
      };
    }

    const tmpPath = `${installPath}.tmp`;
    const totalSize = Number(downloadRes.headers.get("content-length") || 0);
    await downloadWithProgressImpl(downloadRes.body, tmpPath, totalSize, {
      label: asset.name,
    });

    try {
      replaceInstalledJar(tmpPath, installPath);
    } catch (err) {
      const code = (err as NodeJS.ErrnoException)?.code;
      if (code === "EPERM" || code === "EACCES" || code === "EBUSY") {
        return {
          ok: false,
          message:
            `Cannot replace ${installPath}: the file is in use by a running session. ` +
            `Close sessions with 'decx process close --all' and retry.`,
        };
      }
      return { ok: false, message: `Failed to save downloaded file: ${err instanceof Error ? err.message : String(err)}` };
    }

    return {
      ok: true,
      message: `Installed decx-server ${release.tag_name} to ${installPath}`,
      version,
      path: installPath,
    };
  } catch (err) {
    return { ok: false, message: `Installation failed: ${err}` };
  }
}

File: decx-cli/src/installer.rs
//! decx-server.jar discovery, version probing, and installation from GitHub
//! releases.
//!
//! Release discovery avoids the rate-limited GitHub REST API: the latest
//! stable version comes from the npm registry (the npm package is published
//! from the same tag as the server jar) and prereleases from the GitHub
//! releases atom feed. Jar assets use deterministic `/releases/download` URLs.

use std::io::Read;
use std::path::Path;

use crate::error::{DecxError, DecxResult};

const NPM_LATEST_URL: &str = "https://registry.npmjs.org/@jygzyc/decx-cli/latest";
const GITHUB_REPO: &str = "jygzyc/decx";
const RELEASES_ATOM_URL: &str = "https://github.com/jygzyc/decx/releases.atom";

/// Compare two semver strings ("2.2.1" vs "2.3.0"). Returns Ordering-like i32.
pub fn compare_semver(a: &str, b: &str) -> i32 {
    let parse = |s: &str| -> Vec<u64> {
        s.split('-')
            .next()
            .unwrap_or("")
            .split('.')
            .map(|p| p.parse().unwrap_or(0))
            .collect()
    };
    let (pa, pb) = (parse(a), parse(b));
    for i in 0..3 {
        let va = pa.get(i).copied().unwrap_or(0);
        let vb = pb.get(i).copied().unwrap_or(0);
        if va != vb {
            return if va > vb { 1 } else { -1 };
        }
    }
    0
}

pub fn normalize_version(tag: &str) -> String {
    tag.strip_prefix('v').unwrap_or(tag).to_string()
}

fn asset_url(version: &str) -> String {
    format!("https://github.com/{GITHUB_REPO}/releases/download/v{version}/decx-server-{version}.jar")
}

/// Fetch the latest version string: npm registry for stable releases, the
/// GitHub atom feed for prereleases.
pub fn fetch_latest_version(prerelease: bool) -> DecxResult<String> {
    if prerelease {
        let text = crate::net::http_get_text(RELEASES_ATOM_URL, "application/atom+xml")?;
        let is_prerelease_tag = |t: &str| -> bool {
            let v = t.strip_prefix('v').unwrap_or(t);
            v.split('-').count() > 1 && v.split('.').count() >= 3 && v.chars().next().is_some_and(|c| c.is_ascii_digit())
        };
        let mut latest: Option<String> = None;
        for title in extract_tag_titles(&text) {
            if is_prerelease_tag(&title) {
                let version = normalize_version(&title);
                if latest.as_ref().is_none_or(|cur| compare_semver(&version, cur) > 0) {
                    latest = Some(version);
                }
            }
        }
        return latest.ok_or_else(|| DecxError::connection("No prerelease found".to_string()));
    }

    let body = crate::net::http_get_json(NPM_LATEST_URL)?;
    body.get("version")
        .and_then(serde_json::Value::as_str)
        .map(|v| v.to_string())
        .filter(|v| !v.is_empty())
        .ok_or_else(|| DecxError::connection("npm registry returned no version".to_string()))
}

fn extract_tag_titles(atom: &str) -> Vec<String> {
    let mut titles = Vec::new();
    let mut rest = atom;
    while let Some(start) = rest.find("<title>") {
        let after = &rest[start + 7..];
        if let Some(end) = after.find("</title>") {
            titles.push(after[..end].trim().to_string());
            rest = &after[end + 8..];
        } else {
            break;
        }
    }
    titles
}

/// Read `version=<x.y.z>` from the `version.properties` entry inside a jar by
/// parsing the zip central directory directly (stored or deflated entries).
pub fn read_jar_version_property(jar: &Path) -> Option<String> {
    let buf = std::fs::read(jar).ok()?;
    let eocd = find_eocd(&buf)?;
    let entry_count = u16::from_le_bytes([buf[eocd + 10], buf[eocd + 11]]) as usize;
    let mut p = u32::from_le_bytes([buf[eocd + 16], buf[eocd + 17], buf[eocd + 18], buf[eocd + 19]]) as usize;
    for _ in 0..entry_count {
        if buf.get(p..p + 4)? != [0x50, 0x4b, 0x01, 0x02] {
            return None; // central dir signature
        }
        let method = u16::from_le_bytes([buf[p + 10], buf[p + 11]]);
        let compressed_size = u32::from_le_bytes([buf[p + 20], buf[p + 21], buf[p + 22], buf[p + 23]]) as usize;
        let name_len = u16::from_le_bytes([buf[p + 28], buf[p + 29]]) as usize;
        let extra_len = u16::from_le_bytes([buf[p + 30], buf[p + 31]]) as usize;
        let comment_len = u16::from_le_bytes([buf[p + 32], buf[p + 33]]) as usize;
        let local_offset = u32::from_le_bytes([buf[p + 42], buf[p + 43], buf[p + 44], buf[p + 45]]) as usize;
        let name = String::from_utf8_lossy(&buf[p + 46..p + 46 + name_len]).to_string();
        if name == "version.properties" {
            let l_name_len = u16::from_le_bytes([buf[local_offset + 26], buf[local_offset + 27]]) as usize;
            let l_extra_len = u16::from_le_bytes([buf[local_offset + 28], buf[local_offset + 29]]) as usize;
            let data_start = local_offset + 30 + l_name_len + l_extra_len;
            let data = buf.get(data_start..data_start + compressed_size)?;
            let content = match method {
                8 => inflate_raw(data)?,
                _ => String::from_utf8_lossy(data).to_string(),
            };
            for line in content.lines() {
                if let Some(version) = line.strip_prefix("version=") {
                    let version = version.trim();
                    if !version.is_empty() {
                        return Some(version.to_string());
                    }
                }
            }
            return None;
        }
        p += 46 + name_len + extra_len + comment_len;
    }
    None
}

fn find_eocd(buf: &[u8]) -> Option<usize> {
    buf.windows(4)
        .rposition(|w| w == [0x50, 0x4b, 0x05, 0x06])
}

fn inflate_raw(data: &[u8]) -> Option<String> {
    let mut decoder = flate2::read::DeflateDecoder::new(data);
    let mut out = String::new();
    decoder.read_to_string(&mut out).ok()?;
    Some(out)
}

/// Install (or skip-if-current) decx-server.jar. `home` is DECX_HOME.
pub fn install_decx_server(home: &Path, prerelease: bool, current_version: Option<&str>) -> DecxResult<serde_json::Value> {
    eprintln!("  Fetching latest {} info...", if prerelease { "prerelease" } else { "release" });
    let version = fetch_latest_version(prerelease)?;
    let install_dir = home.join("bin");
    let install_path = install_dir.join("decx-server.jar");

    // Prefer the version embedded in the installed jar (ground truth) over the
    // config record, which may be stale after a manual jar replacement.
    let jar_version = install_path.exists().then(|| read_jar_version_property(&install_path)).flatten();
    let effective = jar_version.as_deref().or(current_version);
    if effective == Some(version.as_str()) && install_path.exists() {
        return Ok(serde_json::json!({
            "ok": true,
            "message": format!("decx-server is already up to date (v{version})"),
            "version": version,
            "path": install_path.display().to_string(),
        }));
    }

    std::fs::create_dir_all(&install_dir)
        .map_err(|e| DecxError::internal(format!("cannot create {}: {e}", install_dir.display())))?;
    let url = asset_url(&version);
    eprintln!("  Downloading {url} ...");
    let tmp_path = install_path.with_extension("jar.tmp");
    let downloaded = crate::net::download_to_file(&url, &tmp_path)?;
    if downloaded == 0 {
        let _ = std::fs::remove_file(&tmp_path);
        return Err(DecxError::connection(
            "Download failed: empty response body (the release may still be publishing; retry shortly)".to_string(),
        ));
    }
    replace_installed_jar(&tmp_path, &install_path)?;
    Ok(serde_json::json!({
        "ok": true,
        "message": format!("Installed decx-server v{version} to {}", install_path.display()),
        "version": version,
        "path": install_path.display().to_string(),
    }))
}

fn replace_installed_jar(tmp: &Path, install: &Path) -> DecxResult<()> {
    let backup = install.with_extension("jar.bak");
    let had_existing = install.exists();
    if had_existing {
        let _ = std::fs::rename(install, &backup);
    }
    if let Err(e) = std::fs::rename(tmp, install) {
        // Best-effort rollback before surfacing the failure.
        if had_existing && !install.exists() {
            let _ = std::fs::rename(&backup, install);
        }
        let busy = e.kind() == std::io::ErrorKind::PermissionDenied;
        return Err(DecxError::process(if busy {
            format!(
                "Cannot replace {}: the file may be in use by a running session. \
                 Close sessions with 'decx project close --all' and retry.",
                install.display()
            )
        } else {
            format!("Failed to save downloaded file: {e}")
        }));
    }
    if had_existing {
        let _ = std::fs::remove_file(&backup);
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn semver_compare() {
        assert_eq!(compare_semver("2.2.1", "2.3.0"), -1);
        assert_eq!(compare_semver("2.3.0", "2.3.0"), 0);
        assert_eq!(compare_semver("3.0.0", "2.9.9"), 1);
        assert_eq!(compare_semver("2.2", "2.2.1"), -1);
    }

    #[test]
    fn version_normalization() {
        assert_eq!(normalize_version("v4.2.0"), "4.2.0");
        assert_eq!(normalize_version("4.2.0"), "4.2.0");
    }

    #[test]
    fn atom_titles_extraction() {
        let atom = r#"<feed><entry><title>v4.2.0</title></entry><entry><title>v4.3.0-rc.1</title></entry></feed>"#;
        assert_eq!(extract_tag_titles(atom), vec!["v4.2.0".to_string(), "v4.3.0-rc.1".to_string()]);
    }

    #[test]
    fn jar_version_reader_on_real_zip() {
        // Build a minimal stored-entry zip containing version.properties.
        let content = b"version=4.2.1\n";
        let mut zip: Vec<u8> = Vec::new();
        let name = b"version.properties";
        let crc: u32 = 0; // readers here don't verify crc
        let local_header_offset: u32 = 0;
        // local file header
        zip.extend_from_slice(&[0x50, 0x4b, 0x03, 0x04]);
        zip.extend_from_slice(&[10, 0, 0, 0]); // version, flags
        zip.extend_from_slice(&[0, 0]); // method: stored
        zip.extend_from_slice(&[0, 0, 0, 0]); // time/date
        zip.extend_from_slice(&crc.to_le_bytes());
        zip.extend_from_slice(&(content.len() as u32).to_le_bytes());
        zip.extend_from_slice(&(content.len() as u32).to_le_bytes());
        zip.extend_from_slice(&(name.len() as u16).to_le_bytes());
        zip.extend_from_slice(&[0, 0]); // extra len
        zip.extend_from_slice(name);
        zip.extend_from_slice(content);
        let central_offset = zip.len() as u32;
        // central directory header (fixed layout, method=0 stored)
        zip.extend_from_slice(&[0x50, 0x4b, 0x01, 0x02]); // sig @0
        zip.extend_from_slice(&[10, 0]); // version made by @4
        zip.extend_from_slice(&[10, 0]); // version needed @6
        zip.extend_from_slice(&[0, 0]); // flags @8
        zip.extend_from_slice(&[0, 0]); // method @10 (stored)
        zip.extend_from_slice(&[0, 0]); // mod time @12
        zip.extend_from_slice(&[0, 0]); // mod date @14
        zip.extend_from_slice(&crc.to_le_bytes()); // crc @16
        zip.extend_from_slice(&(content.len() as u32).to_le_bytes()); // csize @20
        zip.extend_from_slice(&(content.len() as u32).to_le_bytes()); // usize @24
        zip.extend_from_slice(&(name.len() as u16).to_le_bytes()); // name len @28
        zip.extend_from_slice(&[0, 0]); // extra len @30
        zip.extend_from_slice(&[0, 0]); // comment len @32
        zip.extend_from_slice(&[0, 0]); // disk @34
        zip.extend_from_slice(&[0, 0]); // internal attrs @36
        zip.extend_from_slice(&[0, 0, 0, 0]); // external attrs @38
        zip.extend_from_slice(&local_header_offset.to_le_bytes()); // offset @42
        zip.extend_from_slice(name);
        let central_size = (zip.len() - central_offset as usize) as u32;
        // EOCD
        zip.extend_from_slice(&[0x50, 0x4b, 0x05, 0x06]);
        zip.extend_from_slice(&[0, 0, 0, 0]);
        zip.extend_from_slice(&1u16.to_le_bytes());
        zip.extend_from_slice(&1u16.to_le_bytes());
        zip.extend_from_slice(&central_size.to_le_bytes());
        zip.extend_from_slice(&central_offset.to_le_bytes());
        zip.extend_from_slice(&[0, 0]);

        let tmp = std::env::temp_dir().join(format!("decx-jar-{}.jar", std::process::id()));
        std::fs::write(&tmp, &zip).unwrap();
        assert_eq!(read_jar_version_property(&tmp).as_deref(), Some("4.2.1"));
        let _ = std::fs::remove_file(&tmp);
    }
}
